### ISSUE TITLE
Camera angle can tank performance

### ISSUE THREAD

**Issue by Triticum0** [xemu-project/xemu#1137]
### Bug Description

Some games with variable camera angles can lose lots of performance depending on where the camera is pointing, This mostly affects sports games and some other outliers.

Games Affected 
https://xemu.app/titles/54540091/#College-Hoops-2K6
https://xemu.app/titles/545400af/#College-Hoops-2K7
https://xemu.app/titles/4b4f0002/#Crimson-Sea
https://xemu.app/titles/53450031/#ESPN-NBA-2K5
https://xemu.app/titles/53450030/#ESPN-NFL-2K5
https://xemu.app/titles/53450037/#Major-League-Baseball-2K5
https://xemu.app/titles/54540096/#Major-League-Baseball-2K5-World-Series-Edition
https://xemu.app/titles/54540092/#Major-League-Baseball-2K6
https://xemu.app/titles/545400b1/#Major-League-Baseball-2K7
https://xemu.app/titles/494c0001/#Metal-Dungeon
https://xemu.app/titles/5454008d/#NHL-2K6
https://xemu.app/titles/545400ad/#NHL-2K7
https://xemu.app/titles/434d0021/#Club-Football
https://xemu.app/titles/434d003f/#Club-Football-2005
https://xemu.app/titles/4541009f/#Madden-NFL-07
https://xemu.app/titles/454100ac/#Madden-NFL-08
https://xemu.app/titles/4541023b/#Madden-NFL-09

Here a video from metal dungon https://youtu.be/CaBnnYwhwP4?t=60
### Expected Behavior

Should be stable performance no mater where the camera pointing

### xemu Version

Version: 0.7.55
Branch: master
Commit: https://github.com/mborgerson/xemu/commit/db389b15086753cb7b72e05b24af205baa742bc3
Date: Sun Jun 26 05:58:57 UTC 2022

### System Information

<html><body>

Field | Value
-- | --
OS | Windows 10
CPU | AMD Ryzen 5 2600 Six-Core Processo
Graphics Device | NVIDIA GeForce RTX 3060 Ti/PCIe/SSE2
Graphics Driver | 4.0.0 NVIDIA 512.95

</body>
</html>

**Comment by RetroPCYT** [comment]
One game more affected for this 'issue": GTA Vice City.

**Comment by MaouYumisu1** [comment]
Used debug build for 06-07 to showcase the camera issue. 07's debug matches retail build. FPS gets worse the further you get from White Tees.
https://user-images.githubusercontent.com/56532164/177433870-e360ddab-1b06-43f3-8f2b-cd307485c1a2.mp4
https://youtu.be/8fkrYXTLp4Y
Affects https://xemu.app/titles/45410021/#Tiger-Woods-PGA-Tour-2003 in specific spots as shown in clip.
https://xemu.app/titles/4541007e/#Tiger-Woods-PGA-Tour-06 06 is very obvious https://youtu.be/55DY0mQF6-c.
https://xemu.app/titles/454100a4/#Tiger-Woods-PGA-Tour-07 Same issue like with 06. https://youtu.be/XstLSi_JFsM
I presume the issues aren't there in 04 and 05. I know they aren't for these two on 0.7.2 for these two games. I'll most likely need to test more on 03 to find more fps drops due to the camera on the latest build.

**Comment by SolusGrimwelder** [comment]
In Crimson Sea which I just tried in latest stable build I noticed what seemed to be a double buffering issue when rotating camera. Framerate dropped straight to 30 when it couldn’t maintain 60. Not sure if this is the issue being described in all of these cases, but I noted it occurred most often when panning the camera. Never played the game on original hardware so I can’t say if this is an emulation issue or if the game was double buffered from the start, but if the latter than the game is in fact performing correctly. Would need a triple buffer patch to address. Will do some more experimenting.

On phone so couldn’t watch linked videos above. Will update post later with any new info.

**Comment by PeruvianPanda** [comment]
I tried out Crimson Sea and here's what I see: some areas / camera angles run at half speed. An easy way to see it is the first level, you spawn in at half speed (30 fps), run forward and around the corner, now you're at full speed (60 fps). Next corner you're back at 30 then around the corner after that you're up to 60 again. Turn around and you can run through this area backwards at 60 but will be at 30 again when you turn back around. The whole first level is similar to this.

I thought it might always be when looking "north" but it's not, it seems to be when you're looking toward your next objective. Going from let's plays on youtube, original hardware doesn't have the slowdown and runs at 60 fps for the whole level.

**Comment by OrganicOstrich** [comment]
World Series Baseball 2K3 can be removed as have tested and no performance issues.

**Comment by riccardobellitto** [comment]
> Would need a triple buffer patch to address. 

I can confirm that enabling the Triple Buffer on the Nvidia Control Panel does improve things, yet sometimes if there's too many enemies or depending on the angle you're facing the boss, the performance drops. I've registered a clip just to show that, hoping will help the devs to fix this issue.

https://github.com/user-attachments/assets/a257ef2e-d388-44e9-b2de-ceca8d5d2e86

**Comment by SolusGrimwelder** [comment]
So let me clarify for the commenters what is happening here and why. What you are seeing is a form of V-Sync designed to eliminate screen tearing. It is not a bug in Xemu but the intended behavior one would expect in a double buffered v-sync. Basically if the engine is unable to render the frame in time it simply drops the frame and begins rendering the next one, hence why the frame rate halves. By comparison a triple buffered v-sync will drop the sync if the frame rate drops below a certain threshold, in this case 60 fps. This introduces tearing but allows for potentially higher average frame rates. 

Long story short, this is not a bug nor something that needs to be addressed by the Xemu team as it is correctly emulated behavior. The issue is that the computer the emulator is running on is not powerful enough. Every time the frame rate drops in a triple buffered game is where you would see that drop to 30 fps in a double buffered game. An option in Xemu to force triple buffering might potentially make things easier but there is no emulation bug occurring here.

**Comment by riccardobellitto** [comment]
> So let me clarify for the commenters what is happening here and why. What you are seeing is a form of V-Sync designed to eliminate screen tearing. It is not a bug in Xemu but the intended behavior one would expect in a double buffered v-sync. Basically if the engine is unable to render the frame in time it simply drops the frame and begins rendering the next one, hence why the frame rate halves. By comparison a triple buffered v-sync will drop the sync if the frame rate drops below a certain threshold, in this case 60 fps. This introduces tearing but allows for potentially higher average frame rates. 
> 
> 
> 
> Long story short, this is not a bug nor something that needs to be addressed by the Xemu team as it is correctly emulated behavior. The issue is that the computer the emulator is running on is not powerful enough. Every time the frame rate drops in a triple buffered game is where you would see that drop to 30 fps in a double buffered game. An option in Xemu to force triple buffering might potentially make things easier but there is no emulation bug occurring here.

I'm sorry, I'm running xemu on a Ryzen 5 5600x with a Rog Strix 3060 12gb and 32gb of RAM @ 3600mhz. Is this build not powerful enough?

Maybe I need to change some settings inside the nvidia control panel or the xemu app, but I don't think the issue is caused by the hardware not being powerful enough.

**Comment by PeruvianPanda** [comment]
> So let me clarify ...
Ok, I see what you mean, it drop to 30 fps right as it gets to 17ms per frame and if the game is hard coded around running at 60 fps then forcing 30 fps also halves the game speed.

I also thought the specifics of when it happens might point to some inefficiency or whatever causing the performance hit.

**Comment by SolusGrimwelder** [comment]
> > So let me clarify ...
> > Ok, I see what you mean, it drop to 30 fps right as it gets to 17ms per frame and if the game is hard coded around running at 60 fps then forcing 30 fps also halves the game speed.
> 
> I also thought the specifics of when it happens might point to some inefficiency or whatever causing the performance hit.

Response:

Exactly this. If the game logic is tied to the framerate, something we know is not unheard of for Japanese console game developers from this era then a halving of the framerate would also halve the world simulation, causing the "bug" we are seeing. The halving of the framerate is occurring as said exactly when the framerate exceeds the 17ms per frame target of a 60 fps game, indicating the game is double buffered, meaning it either hits its frame target or it simply waits for the next frame to render it. There is no half frames rendered in a double buffered setup. 

That is what I mean when I say the behavior is not a bug with the emulation. Now that does not mean that the emulation could not be optimized to avoid the frame drop to begin with, but the game logic halving in speed is likely not a bug. For a more modern example of a similar issue, remember the challenges of running Dark Souls at 60 FPS because the physics and damage calculations were run at the render rate? Japanese developers from this era were not targeting PC hardware. This was an Xbox exclusive game where they could carefully control the framerate to ensure it never varied from their target. That's why it was likely designed this way. 

__

Update:

I updated my computer to use a Ryzen 7 9800x3D paired with 32 GB of High speed DDR5 RAM. Haven't been able to get one of the new graphics cards yet and will update when I do but even without an updated graphics card I can confirm that the issue was all but eliminated on my new computer even when running at 3x render resolution. Had a couple drops when making sudden camera changes and one during the boss fight at the end of the tutorial when he fired some particles off but this could be due to the game currently being stored on an old mechanical drive. I'll put it onto my new Samsung 990 SSD to see if that alleviates the remaining issues. But even in the final room of the tutorial with a dozen enemies on screen and me spinning the camera around the game never dropped a frame. Only when quickly facing the camera to a newly loaded asset did I see a drop.

**Comment by Triticum0** [comment]
Retested no change

**Comment by SolusGrimwelder** [comment]
I just tested Crimson Sea again using my latest build which is running a Ryzen 7 9800x3D, Geforce 5070 ti, and 32 GB of RAM. The game is stored on a Samsung 990 SSD as is the emulator itself. Utilizing these specs and running at 4x native resolution still saw the occasional performance drop to 30 fps that one would expect from a game with non-adaptive double buffered V-Sync enabled. I noticed that at worst I was hitting about 19 ms of frame time, significantly better than the close to 30 on my old computer, but in this rendering setup, it makes little difference how much you exceed the target frame time, simply that you do exceed it. So I tried switching to OpenGL to see if it could potentially run the emulation ever so slightly faster, and sure enough, that did it. My frame time dropped to 16 ms at worst, meaning I saw zero frame drops at all during my run of play. I later streamed myself playing and the extra overhead of running OBS was enough to cause the occasional bout of 17 ms gameplay which, again, was indistinguishable from 30 ms gameplay. 

Long story short, this has confirmed my belief that at least in the case of this specific game, there seems to be no specific bug causing the frame drop. The difference between flawless emulation and a laggy mess is 1 ms of frame time. That type of difference is to be expected in almost any game from scene to scene. The issue is simply that due to the very antiquated way this engine is designed, those normal small fluctuations cause abnormally large framerate drops. Now of course the emulator could likely be optimized overall to try and maintain a higher framerate on a larger number of machines, but I reiterate that in the case of this game at least, there is no specific bug causing dramatic performance variations, just an antiquated rendering setup performing how it is supposed to. I will test some of the other games tagged with this issue and report back, but my strong guess is that in almost all cases turning the camera is changing the rendering load just enough to trigger a framerate drop and with the game logic tied to framerate, that causes the issues we are seeing.

**Comment by blakeplusplus** [comment]
Tested Madden 07 on OpenGL, with both 1x and 2x resolution scaling.  Performance is actually very steady besides occasional blips that last a quarter of a second. 
On Vulkan, it's atrocious and unplayable.

```
Version:      0.8.71
Branch:       master
Commit:       e02e41ccaaceffcad816cfb554dd1195767e952d
Date:         Fri Jun  6 03:00:46 AM UTC 2025

CPU:          AMD Ryzen 7 7800X3D 8-Core Processor           
OS Platform:  Linux
OS Version:   Freedesktop SDK 24.08 (Flatpak runtime)
Manufacturer: AMD
GPU Model:    AMD Radeon RX 7900 XT (radeonsi, navi31, LLVM 19.1.7, DRM 3.63, 6.15.2-arch1-1)
Driver:       4.6 (Core Profile) Mesa 25.1.1 (git-7485541cc3)
Shader:       4.60
```
